src/map/script.cpp

```diff
@@ -16064,49 +16064,106 @@ BUILDIN_FUNC(chatmes)
 	return SCRIPT_CMD_SUCCESS;
 }
 
-// change npc walkspeed [Valaris]
+/**
+ * Change npc walkspeed.
+ * npcspeed <speed value> {,"<npc name>"};
+ */
 BUILDIN_FUNC(npcspeed)
 {
-	struct npc_data* nd;
-	int speed;
+	npc_data *nd;
 
-	speed = script_getnum(st,2);
-	nd =(struct npc_data *)map_id2bl(st->oid);
+	if (script_hasdata(st, 3))
+		nd = npc_name2id(script_getstr(st, 3));
+	else
+		nd = map_id2nd(st->oid);
 
-	if( nd ) {
-		nd->speed = speed;
-		nd->ud.state.speed_changed = 1;
+	if (!nd) {
+		if (script_hasdata(st, 3))
+			ShowError("buildin_npcspeed: %s is a non-existing NPC.\n", script_getstr(st, 3));
+		else
+			ShowError("buildin_npcspeed: non-existing NPC.\n");
+		return SCRIPT_CMD_FAILURE;
 	}
 
+	int speed = script_getnum(st, 2);
+	
+	if (speed < MIN_WALK_SPEED || speed > MAX_WALK_SPEED) {
+		ShowError("buildin_npcspeed: invalid speed %d (min: %d, max: %d).\n", speed, MIN_WALK_SPEED, MAX_WALK_SPEED);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	nd->speed = speed;
+	nd->ud.state.speed_changed = 1;
+
 	return SCRIPT_CMD_SUCCESS;
 }
-// make an npc walk to a position [Valaris]
+
+/**
+ * Make an npc walk to a position.
+ * npcwalkto <x>,<y> {,"<speed>" {,"<npc name>"} };
+ */
 BUILDIN_FUNC(npcwalkto)
 {
-	struct npc_data *nd=(struct npc_data *)map_id2bl(st->oid);
-	int x=0,y=0;
+	npc_data *nd;
 
-	x=script_getnum(st,2);
-	y=script_getnum(st,3);
+	if (script_hasdata(st, 5))
+		nd = npc_name2id(script_getstr(st, 5));
+	else
+		nd = map_id2nd(st->oid);
 
-	if(nd) {
-		if (!nd->status.hp)
-			status_calc_npc(nd, SCO_FIRST);
+	if (!nd) {
+		if (script_hasdata(st, 5))
+			ShowError("buildin_npcwalkto: %s is a non-existing NPC.\n", script_getstr(st, 5));
 		else
-			status_calc_npc(nd, SCO_NONE);
-		unit_walktoxy(&nd->bl,x,y,0);
+			ShowError("buildin_npcwalkto: non-existing NPC.\n");
+		return SCRIPT_CMD_FAILURE;
 	}
+
+	if (script_hasdata(st, 4)) {
+		int speed = script_getnum(st, 4);
+		
+		if (speed < MIN_WALK_SPEED || speed > MAX_WALK_SPEED) {
+			ShowError("buildin_npcwalkto: invalid speed %d (min: %d, max: %d).\n", speed, MIN_WALK_SPEED, MAX_WALK_SPEED);
+			return SCRIPT_CMD_FAILURE;
+		}
+		nd->speed = speed;
+		nd->ud.state.speed_changed = 1;
+	}
```

Comment:
Since this affects the speed of the NPC after any call to this function:
Why do we have to make it a parameter for this function and why can npcspeed not be called by the user before calling this function?

Reply:
The "speed" parameter was to match aegis command. Maybe this parameter is overkill since we save the walking speed indefinitely. Ideally I would prefer to have it.

Should I remove the parameter?
How about
```
	if (nd->speed != speed) {
		if (speed < MIN_WALK_SPEED || speed > MAX_WALK_SPEED) {
			ShowError("buildin_npcwalkto: invalid speed %d (min: %d, max: %d).\n", speed, MIN_WALK_SPEED, MAX_WALK_SPEED);
		etc..
	}
```

Reply:
If you prefer to have it, it is ok.
But we should document it somehow, because I think for the users it creates the impression that the speed will only be changed for "this" movement and not forever.
Thats why I would have removed it to not create confusion.

Reply:
The speed parameter was finally removed to avoid confusion